### PR TITLE
UI: Card and Deck

### PR DESCRIPTION
**Pull request by Amstutz**
This pull request is designed to conform with the new guidelines to add new Components to the centralized UI-components of ILIAS. It is to be discussed in the JF. Note that this PR is not holding the final implementation of the new component, it only describes it's interfaces and gives an example implementation of the two components.

As a short reminder, for new UI-Components in ILIAS the following rules need to apply:
- New methods on factory interfaces of the framework MUST be agreed upon by the
  Jour Fixe, as new methods on factories represent new abstract nodes in the Kitchen
  Sink taxonomy or new concrete UI components.
- To propose a new UI component a pull request against the trunk MUST be made on
  github. The code in the pull request SHOULD obay the rules given in **Interfaces
  to Factories** and **Interfaces to UI components**. The existing unit tests for
  the UI framework SHOULD pass.
- The pull request MAY be made from the edge branch in the ILIAS-repo. If the new
  component is already implemented, the edge installation of ILIAS MAY be used
  for showcasing the component.
- The new method MUST be backed with a stub implementation down to the methods
  that represent concrete UI components, where said methods MUST raise
  ILIAS\UI\NotImplementedException upon call, if the UI component is not already
  implemented.
- The proposed UI component MAY already be implemented. If the UI component is
  implemented, it SHOULD obay the rules given in **Rules for Implementors**.
- In addition to the YAML-Block described in **Interfaces to Factories** the
  proposed interfaces, if not already implemented, SHOULD contain the following
  fields:
  - `less` - lists the LESS-variables that will be used to render the
    UI component including their purpose
- If the new UI component is not implemented, there SHOULD be an html-example in
  the examples-folder.
- The new UI component MUST be presented on the JF, including the corresponding
  pull request. This SHOULD include some visible representation of the presented
  UI component, like a mock up or a basic implementation on the edge installation.

### CODE REVIEW

File: src/UI/Component/Card/Card.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+namespace ILIAS\UI\Component\Card;
+
+/**
+ * Interface Card
+ * @package ILIAS\UI\Component\Card
+ */
+interface Card extends \ILIAS\UI\Component\Component {
+
+	/**
+     * Sets the title in the heading section of the card
```
Comment (klees) [inline src/UI/Component/Card/Card.php]: Looks as if you inherited some bad indentation from me =)

File: src/UI/Component/Card/Card.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	 * @return string
+	 */
+	public function getTitle();
+
+	/**
+     * Set additional content displayed directly underneath the title
+	 * @param \ILIAS\UI\Component\Component $section
+	 * @return Card
+	 */
+	public function withHeaderSection(\ILIAS\UI\Component\Component $section);
```
Comment (klees) [inline src/UI/Component/Card/Card.php]: Why not withHeader?

File: src/UI/Component/Card/Card.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	 * @return \ILIAS\UI\Component\Component
+	 */
+	public function getHeaderSection();
+
+	/**
+     * Set multiple sections of the card as array
+	 * @param \ILIAS\UI\Component\Component[] $sections
+	 * @return Card
+	 */
+	public function withContentSections($sections);
```
Comment (klees) [inline src/UI/Component/Card/Card.php]: Why not withContents?

Comment (klees) [inline src/UI/Component/Card/Card.php]: You may also hint on array.

File: src/UI/Factory.php [hunk 4]
```diff
@@ -162,4 +162,66 @@ public function glyph();
 	 * @return  \ILIAS\UI\Component\Button\Factory
 	 */
 	public function button();
+
+    /**
+     * ---
+     * description:
+     *   purpose: >
+     *     	A card is a flexible content container for small chunks of structured data.
```
Comment (klees) [inline src/UI/Factory.php]: Indentation.

File: src/UI/Factory.php [hunk 5]
```diff
@@ -162,4 +162,66 @@ public function glyph();
...
+     * rules:
+     *   usage:
+     *      1: Decks MUST only be used to display multiple Cards.
+     *   style:
+     *      1: The number of cards displayed per row MUST adapt to the screen size.
+     * ---
+     * @param \ILIAS\UI\Component\Card\Card[] $cards
+     * @return \ILIAS\UI\Component\Deck\Deck
+     */
+    public function deck($cards);
```
Comment (klees) [inline src/UI/Factory.php]: You may hint on array.

File: src/UI/Factory.php [hunk 6]
```diff
@@ -162,4 +162,66 @@ public function glyph();
...
+     *      1: Cards MUST contain a title.
+     *      2: Cards SHOULD contain an Image or Icon in the header section.
+     *      3: Cards MAY contain Interaction Triggers.
+     *   style:
+     *      1: Sections of  Cards MUST be separated by Dividers.
+     *   accessibility:
+     *      1: If multiple Cards are used, they MUST be contained in a Deck.
+     * ---
+     * @param string $title
+     * @param \ILIAS\UI\Component\Image\Image $image
```
Comment (klees) [inline src/UI/Factory.php]: Where can i get one of those?

File: src/UI/Implementation/Component/Card/Card.php [hunk 7]
```diff
@@ -0,0 +1,113 @@
+<?php
```
Comment (klees) [inline src/UI/Implementation/Component/Card/Card.php:1]: The whole file uses spaces for indentation. Please switch to tabs.

File: src/UI/Implementation/Component/Deck/Deck.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
+<?php
```
Comment (klees) [inline src/UI/Implementation/Component/Deck/Deck.php:1]: Tabs instead of spaces please.

File: src/UI/Factory.php [hunk 9]
```diff
@@ -162,4 +162,66 @@ public function glyph();
...
+     * @return \ILIAS\UI\Component\Card\Card
+     */
+    public function card($title, \ILIAS\UI\Component\Image\Image $image = null);
+
+    /**
+     * ---
+     * description:
+     *   purpose: >
+     *     	Decks are used to display multiple Cards in a grid.
```
Comment (klees) [inline src/UI/Factory.php]: Indentation.

File: src/UI/examples/Card/base.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
+<?php
```
Comment (klees) [inline src/UI/examples/Card/base.php:1]: Tabs instead of spaces please.

File: src/UI/examples/Card/with_sections.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
+<?php
```
Comment (klees) [inline src/UI/examples/Card/with_sections.php:1]: Tabs instead of spaces please.

File: src/UI/examples/Deck/base.php [hunk 12]
```diff
@@ -0,0 +1,33 @@
+<?php
```
Comment (klees) [inline src/UI/examples/Deck/base.php:1]: Tabs instead of spaces please.

File: src/UI/examples/Deck/user.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
+<?php
```
Comment (klees) [inline src/UI/examples/Deck/user.php:1]: Tabs instead of spaces please.

File: src/UI/examples/Deck/xl_cards.php [hunk 14]
```diff
@@ -0,0 +1,33 @@
+<?php
```
Comment (klees) [inline src/UI/examples/Deck/xl_cards.php]: Tabs instead of spaces please.

File: tests/UI/Base.php [hunk 15]
```diff
@@ -21,6 +21,8 @@ class NoUIFactory implements Factory {
 	public function counter() {}
 	public function glyph() {}
 	public function button() {}
+    public function deck($cards) {}
```
Comment (klees) [inline tests/UI/Base.php]: Indentation.

File: tests/UI/MainFactoryTest.php [hunk 16]
```diff
@@ -3,7 +3,10 @@
 require_once 'tests/UI/AbstractFactoryTest.php';
 
 class MainFactoryTest extends AbstractFactoryTest {
-	public $kitchensink_info_settings = array();
+	public $kitchensink_info_settings = array
+    ( "card"		=> array("context" => false)
```
Comment (klees) [inline tests/UI/MainFactoryTest.php]: Indentation.

File: src/UI/Component/Card/Card.php [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+/* Copyright (c) 2016 Timon Amstutz <timon.amstutz@ilub.unibe.ch> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Component\Card;
+
+/**
+ * Interface Card
+ * @package ILIAS\UI\Component\Card
+ */
+interface Card extends \ILIAS\UI\Component\Component {
```
Comment (klees) [inline src/UI/Component/Card/Card.php:11]: Does this model fit the description in the Factory? Looks as if i could have a title, an image and a header section. This seems to be odd, or i do not understand the description correctly. Can the header section contain another title and image? How would that be rendered without looking weird?

**Comment by matthiaskunkel** [comment]
Jour Fixe, August 01, 2016: We appreciate the suggestion and schedule these new UI elements for ILIAS 5.2.

**Comment by Amstutz** [comment]
Hi Richard, thank you for your feedback.

I added 4 commits following your recommendations. Is this nasty intention problem now out of the world?

**Pull request merged**